ExplicitTypeInterfaceRule rule

kimdv opened the pull request:
Hello! 

I don't want to start a war here, so I made this an opt in as default!

To explain myself briefly; I added this rule because I think it makes the code more readable and helps keep an overview of the specific properties. 

As stated before - the rule is made an OPT IN, so people themselves can decide whether to follow the rule or not. 

I have yet to figure out how the test works with the `nonTriggeringExamples`, `triggeringExamples` and `↓`. 

I hope that there is a nice person, who will help me out here, so we can finish the rule and give the people who like it, an opportunity to start using it. :)

Remember to keep it civil. 👍

marcelofabri commented on Source/SwiftLintFramework/Rules/TypeInterfaceRule.swift:
```diff
@@ -0,0 +1,58 @@
+
+        // Check if the property have a type
+        if dictionary.typeName != nil {
+                return []
```
too much indentation

marcelofabri commented:
What do you think about renaming it to `ExplicitTypeInterfaceRule` or something like that? Mainly because in the future we may want to have a rule to enforce the opposite.

Also, is there a reason why other kind of variables aren't validated? IMO at least class/static variables should be validate too.

kimdv commented:
@marcelofabri I have changed the name to `ExplicitTypeInterfaceRule ` and added more var types. 

But I still don't know how the tests work, can you please help me out ? :)

marcelofabri commented:
> But I still don't know how the tests work, can you please help me out ? :)

I think you've figured it out, since the tests are now passing 😅 

Can you rebase this to fix the conflicts?

kimdv commented:
hmm, it seems it fails the checks?

Is the tests made wrong? :)


I also resolved the conflicts, so it can be merged with the master.

marcelofabri commented:
@kimdv I haven't checked it locally, but my guess is that you need to wrap `class var` and `static var` declarations instead of a class (or a type for `static`).

You can run the tests on Xcode to debug it further.

kimdv commented:
@marcelofabri I see that all the tests fails?

```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
var ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
let ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
static var ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
class var ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
var ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
let ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
static var ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
class var ↓myVar = 0
```
```

marcelofabri commented:
@kimdv probably for the same reason: if you don't wrap those into a type, they'll be `.varGlobal`, which the rule doesn't handle.

marcelofabri commented:
Can you pleas explain or tell me where I can read about the `↓` and how it works? :)

This marker is used to indicate where the violation should be reported (instead of only validating that a violation is reported).

For some reason the tests were falling on Travis even tough they succeeded on my machine. I've pushed d11b24d which should fix the only failure that happened here.

marcelofabri commented on Source/SwiftLintFramework/Models/MasterRuleList.swift:
```diff
@@ -143,6 +143,7 @@ public let masterRuleList = RuleList(rules:
     TrailingSemicolonRule.self,
     TrailingWhitespaceRule.self,
     TypeBodyLengthRule.self,
+    ExplicitTypeInterfaceRule.self,
```
this should be kept ordered

marcelofabri commented on Tests/SwiftLintFrameworkTests/RulesTests.swift:
```diff
@@ -419,7 +423,8 @@ extension RulesTests {
             ("testVoidReturn", testVoidReturn),
             ("testSuperCall", testSuperCall),
-            ("testWeakDelegate", testWeakDelegate)
+            ("testWeakDelegate", testWeakDelegate),
```
this should be kept ordered

marcelofabri commented on Tests/SwiftLintFrameworkTests/RulesTests.swift:
```diff
@@ -346,6 +346,10 @@ class RulesTests: XCTestCase {
         verifyRule(WeakDelegateRule.description)
     }
 
+    func testExplicitTypeInterfaceRule() {
```
this should be kept ordered

kimdv commented:
@marcelofabri The test and rules should be ordered now. 

Lets see if the test passes now 😃

kimdv commented:
hmm, still one test failing...??

marcelofabri commented:
@kimdv Actually the tests were failing when using SwiftLint with Swift 2.3. I took the liberty to fix them and do some small refactorings on your branch, if that's ok.

@jpsim Since it's a new rule, would you like to review this as well?

marcelofabri commented:
Actually from some reason I can't push to your fork from this computer, so I got tired of fighting against git and opened #1356

marcelofabri commented:
Closing this in favor of #1356
